### ISSUE TITLE
Type "privacy" does not exist

### ISSUE THREAD

**Issue by canadaduane** [omarroth/invidious#926]
On first attempt at `docker-compose up` I received the following error:

```ERROR:  type "privacy" does not exist```

Here is the docker output and surrounding context:

```postgres_1   | ### importing table schemas
postgres_1   | /var/run/postgresql:5432 - accepting connections
postgres_1   | CREATE ROLE
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | 2019-11-21 21:42:39.903 UTC [172] ERROR:  type "privacy" does not exist at character 200
postgres_1   | 2019-11-21 21:42:39.903 UTC [172] STATEMENT:  CREATE TABLE public.playlists
postgres_1   | 	(
postgres_1   | 	    title text,
postgres_1   | 	    id text primary key,
postgres_1   | 	    author text,
postgres_1   | 	    description text,
postgres_1   | 	    video_count integer,
postgres_1   | 	    created timestamptz,
postgres_1   | 	    updated timestamptz,
postgres_1   | 	    privacy privacy,
postgres_1   | 	    index int8[]
postgres_1   | 	);
postgres_1   | ERROR:  type "privacy" does not exist
postgres_1   | LINE 10:     privacy privacy,
postgres_1   |                      ^
postgres_1   | 2019-11-21 21:42:39.903 UTC [172] ERROR:  relation "public.playlists" does not exist
postgres_1   | 2019-11-21 21:42:39.903 UTC [172] STATEMENT:  GRANT ALL ON public.playlists TO kemal;
postgres_1   | ERROR:  relation "public.playlists" does not exist
postgres_1   | 2019-11-21 21:42:39.960 UTC [183] ERROR:  relation "playlists" does not exist
postgres_1   | 2019-11-21 21:42:39.960 UTC [183] STATEMENT:  CREATE TABLE playlist_videos
postgres_1   | 	(
postgres_1   | 	    title text,
postgres_1   | 	    id text,
postgres_1   | 	    author text,
postgres_1   | 	    ucid text,
postgres_1   | 	    length_seconds integer,
postgres_1   | 	    published timestamptz,
postgres_1   | 	    plid text references playlists(id),
postgres_1   | 	    index int8,
postgres_1   | 	    live_now boolean,
postgres_1   | 	    PRIMARY KEY (index,plid)
postgres_1   | 	);
postgres_1   | ERROR:  relation "playlists" does not exist
postgres_1   | 2019-11-21 21:42:39.962 UTC [183] ERROR:  relation "public.playlist_videos" does not exist
postgres_1   | 2019-11-21 21:42:39.962 UTC [183] STATEMENT:  GRANT ALL ON TABLE public.playlist_videos TO kemal;
postgres_1   | ERROR:  relation "public.playlist_videos" does not exist
postgres_1   | CREATE TYPE
postgres_1   | ### invidious database setup finished
```

I suspect this may be a cascade of errors starting here (earlier messages indicating errors connecting):

```
postgres_1   | 2019-11-21 21:42:30.280 UTC [51] LOG:  shutting down
postgres_1   | 2019-11-21 21:42:30.295 UTC [49] LOG:  database system is shut down
postgres_1   |  done
postgres_1   | server stopped
postgres_1   |
postgres_1   | PostgreSQL init process complete; ready for start up.
postgres_1   |
postgres_1   | 2019-11-21 21:42:30.383 UTC [7] LOG:  listening on IPv4 address "0.0.0.0", port 5432
postgres_1   | 2019-11-21 21:42:30.383 UTC [7] LOG:  listening on IPv6 address "::", port 5432
postgres_1   | 2019-11-21 21:42:30.386 UTC [7] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
postgres_1   | 2019-11-21 21:42:30.398 UTC [58] LOG:  database system was shut down at 2019-11-21 21:42:30 UTC
postgres_1   | 2019-11-21 21:42:30.403 UTC [7] LOG:  database system is ready to accept connections
invidious_1  | Unhandled exception:  (DB::ConnectionRefused)
invidious_1  |   from lib/sqlite3/src/sqlite3/connection.cr:8:11 in '->'
invidious_1  |   from lib/db/src/db/pool.cr:255:3 in 'build_resource'
invidious_1  |   from lib/db/src/db/pool.cr:380:11 in 'build_database'
invidious_1  |   from /usr/lib/crystal/core/slice.cr:241:13 in '???'
invidious_1  |   from /usr/lib/crystal/core/crystal/once.cr:255:3 in '__crystal_once'
invidious_1  |   from src/invidious.cr:44:1 in '__crystal_main'
invidious_1  |   from /usr/lib/crystal/core/crystal/main.cr:47:14 in 'main'
invidious_1  |   from ???
invidious_1  | Caused by: Cannot establish connection (PQ::ConnectionError)
invidious_1  |   from lib/pg/src/pq/connection.cr:34:9 in 'new'
invidious_1  |   from lib/pg/src/pg/connection.cr:13:23 in '->'
invidious_1  |   from lib/db/src/db/pool.cr:255:3 in 'build_resource'
invidious_1  |   from lib/db/src/db/pool.cr:380:11 in 'build_database'
invidious_1  |   from /usr/lib/crystal/core/slice.cr:241:13 in '???'
invidious_1  |   from /usr/lib/crystal/core/crystal/once.cr:255:3 in '__crystal_once'
invidious_1  |   from src/invidious.cr:44:1 in '__crystal_main'
invidious_1  |   from /usr/lib/crystal/core/crystal/main.cr:47:14 in 'main'
invidious_1  |   from ???
invidious_1  | Caused by: Error connecting to 'postgres:5432': Connection refused (Errno)
invidious_1  |   from /usr/lib/crystal/core/socket.cr:70:11 in 'new'
invidious_1  |   from lib/pg/src/pg/connection.cr:13:23 in '->'
invidious_1  |   from lib/db/src/db/pool.cr:255:3 in 'build_resource'
invidious_1  |   from lib/db/src/db/pool.cr:380:11 in 'build_database'
invidious_1  |   from /usr/lib/crystal/core/slice.cr:241:13 in '???'
invidious_1  |   from /usr/lib/crystal/core/crystal/once.cr:255:3 in '__crystal_once'
invidious_1  |   from src/invidious.cr:44:1 in '__crystal_main'
invidious_1  |   from /usr/lib/crystal/core/crystal/main.cr:47:14 in 'main'
invidious_1  |   from ???
postgres_1   | 2019-11-21 21:42:33.464 UTC [65] FATAL:  role "kemal" does not exist
invidious_1  | Unhandled exception:  (DB::ConnectionRefused)
invidious_1  |   from lib/sqlite3/src/sqlite3/connection.cr:8:11 in '->'
invidious_1  |   from lib/db/src/db/pool.cr:255:3 in 'build_resource'
invidious_1  |   from lib/db/src/db/pool.cr:380:11 in 'build_database'
invidious_1  |   from /usr/lib/crystal/core/slice.cr:241:13 in '???'
invidious_1  |   from /usr/lib/crystal/core/crystal/once.cr:255:3 in '__crystal_once'
invidious_1  |   from src/invidious.cr:44:1 in '__crystal_main'
invidious_1  |   from /usr/lib/crystal/core/crystal/main.cr:47:14 in 'main'
invidious_1  |   from ???
invidious_1  | Caused by: role "kemal" does not exist (PQ::PQError)
invidious_1  |   from lib/pg/src/pq/connection.cr:162:7 in 'read'
invidious_1  |   from lib/pg/src/pq/connection.cr:157:7 in 'connect'
invidious_1  |   from lib/db/src/db/database.cr:57:9 in '->'
invidious_1  |   from lib/db/src/db/pool.cr:255:3 in 'build_resource'
invidious_1  |   from lib/db/src/db/pool.cr:380:11 in 'build_database'
invidious_1  |   from /usr/lib/crystal/core/slice.cr:241:13 in '???'
invidious_1  |   from /usr/lib/crystal/core/crystal/once.cr:255:3 in '__crystal_once'
invidious_1  |   from src/invidious.cr:44:1 in '__crystal_main'
invidious_1  |   from /usr/lib/crystal/core/crystal/main.cr:47:14 in 'main'
invidious_1  |   from ???
invidious_invidious_1 exited with code 1
postgres_1   | 2019-11-21 21:42:37.409 UTC [66] FATAL:  role "kemal" does not exist
invidious_1  | Unhandled exception:  (DB::ConnectionRefused)
invidious_1  |   from lib/sqlite3/src/sqlite3/connection.cr:8:11 in '->'
invidious_1  |   from lib/db/src/db/pool.cr:255:3 in 'build_resource'
invidious_1  |   from lib/db/src/db/pool.cr:380:11 in 'build_database'
invidious_1  |   from /usr/lib/crystal/core/slice.cr:241:13 in '???'
invidious_1  |   from /usr/lib/crystal/core/crystal/once.cr:255:3 in '__crystal_once'
invidious_1  |   from src/invidious.cr:44:1 in '__crystal_main'
invidious_1  |   from /usr/lib/crystal/core/crystal/main.cr:47:14 in 'main'
invidious_1  |   from ???
invidious_1  | Caused by: role "kemal" does not exist (PQ::PQError)
invidious_1  |   from lib/pg/src/pq/connection.cr:162:7 in 'read'
invidious_1  |   from lib/pg/src/pq/connection.cr:157:7 in 'connect'
invidious_1  |   from lib/db/src/db/database.cr:57:9 in '->'
invidious_1  |   from lib/db/src/db/pool.cr:255:3 in 'build_resource'
invidious_1  |   from lib/db/src/db/pool.cr:380:11 in 'build_database'
invidious_1  |   from /usr/lib/crystal/core/slice.cr:241:13 in '???'
invidious_1  |   from /usr/lib/crystal/core/crystal/once.cr:255:3 in '__crystal_once'
invidious_1  |   from src/invidious.cr:44:1 in '__crystal_main'
invidious_1  |   from /usr/lib/crystal/core/crystal/main.cr:47:14 in 'main'
invidious_1  |   from ???
```

**Comment by charles-m-knox** [comment]
I may have run into this too. Turns out that the latest code for invidious didn't create a few tables in Postgres. These are taken from the README. Connect to your invidious Postgresql instance (may require opening ports) and run these - not necessarily in this order:

```
$ psql invidious kemal < /home/invidious/invidious/config/sql/playlist_videos.sql
$ psql invidious kemal < /home/invidious/invidious/config/sql/privacy.sql
$ psql invidious kemal < /home/invidious/invidious/config/sql/playlists.sql
```

**Comment by Perflyst** [comment]
@canadaduane did it work?

**Comment by canadaduane** [comment]
I eventually figured out how to run these SQL scripts inside docker:

```
docker-compose run postgres -d
docker exec -i invidious_postgres_run_55f415dbd530 /usr/bin/psql invidious kemal <config/sql/playlists.sql
docker exec -i invidious_postgres_run_55f415dbd530 /usr/bin/psql invidious kemal <config/sql/playlist_videos.sql
```

Apparently the `privacy.sql` was already installed.

Unfortunately, this leads to new errors, this time Crystal segfaults:

```docker-compose up
WARNING: The Docker Engine you're using is running in swarm mode.

Compose does not use swarm mode to deploy services to multiple nodes in a swarm. All containers will be scheduled on the current node.

To deploy your application across the swarm, use `docker stack deploy`.

Starting invidious_postgres_1 ... done
Starting invidious_invidious_1 ... done
Attaching to invidious_postgres_1, invidious_invidious_1
postgres_1   | running postgres /usr/local/bin/docker-entrypoint.sh postgres
postgres_1   |
postgres_1   | PostgreSQL Database directory appears to contain a database; Skipping initialization
postgres_1   |
postgres_1   | 2019-11-27 18:12:49.303 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
postgres_1   | 2019-11-27 18:12:49.303 UTC [1] LOG:  listening on IPv6 address "::", port 5432
postgres_1   | 2019-11-27 18:12:49.308 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
postgres_1   | 2019-11-27 18:12:49.330 UTC [26] LOG:  database system was shut down at 2019-11-27 18:12:41 UTC
postgres_1   | 2019-11-27 18:12:49.336 UTC [1] LOG:  database system is ready to accept connections
invidious_1  | Invalid memory access (signal 11) at address 0x10000001b
invidious_1  | [0x55c831f92b86] ???
invidious_1  | [0x55c831a3e1af] __crystal_sigfault_handler +479
invidious_1  | [0x7f25d9c4a28f] ???
invidious_1  | Invalid memory access (signal 11) at address 0x10000001b
invidious_1  | [0x561bd0c6fb86] ???
invidious_1  | [0x561bd071b1af] __crystal_sigfault_handler +479
invidious_1  | [0x7f8f1b36a28f] ???
invidious_invidious_1 exited with code 11
invidious_1  | Invalid memory access (signal 11) at address 0x10000001b
invidious_1  | [0x564f1d59bb86] ???
invidious_1  | [0x564f1d0471af] __crystal_sigfault_handler +479
invidious_1  | [0x7fe7b887228f] ???
invidious_invidious_1 exited with code 11
^CERROR: Aborting.```

**Comment by Perflyst** [comment]
Glad you fixed it. For your new issue, see #917

**Comment by edgd1er** [comment]
I had the same error and I've been searching for a while.
Privacy table is referenced in playlist but created after, If you move in docker/entrypoint.postgres.sh L24 to line 15, after create DB.

**Comment by omarroth** [comment]
Fixed by #1042.
